File: 1570-final-prices-with-a-special-discount-in-a-shop/1570-final-prices-with-a-special-discount-in-a-shop.js
var finalPrices = function(prices) {
    let n = prices.length;
    let ans = [...prices];
    let stack = [];

    for (let i = 0; i < n; i++) {
        while (stack.length > 0 && prices[stack[stack.length - 1]] >= prices[i]) {
            ans[stack.pop()] -= prices[i];
        }
        stack.push(i);
    }

    return ans;
};

File: 1570-final-prices-with-a-special-discount-in-a-shop/1570-final-prices-with-a-special-discount-in-a-shop.rb
def final_prices(prices)
    n = prices.size
    ans = prices.clone
    stack = []

    (0...n).each do |i|
        while !stack.empty? && prices[stack.last] >= prices[i]
            ans[stack.pop] -= prices[i]
        end
        stack.push(i)
    end

    ans
end
